javalib/src/main/scala/java/net/URLDecoder.scala

```diff
@@ -0,0 +1,46 @@
+package java.net
+
+import java.io.UnsupportedEncodingException
+import java.nio.{CharBuffer, ByteBuffer}
+import java.nio.charset.Charset
+
+object URLDecoder {
+  def decode(s: String, enc: String): String = {
+
+    if (!Charset.isSupported(enc)) throw new UnsupportedEncodingException(enc)
+
+    val len = s.length
+    val charsetDecoder = Charset.forName(enc).newDecoder()
+    val charBuffer = CharBuffer.allocate(len)
+
+    var i = 0
+    while (i < len) {
+      s.charAt(i) match {
+        case '+' =>
+          charBuffer.append(' ')
+          i += 1
+
+        case '%' if i + 3 < len =>
+          val chunks = s.substring(i).grouped(3).takeWhile(_.charAt(0) == '%')
+          try {
+            val hexValues = chunks.map(_.substring(1))
+            val bytes = hexValues.map(Integer.parseInt(_, 16).toByte).toArray
+            val decoded = charsetDecoder.decode(ByteBuffer.wrap(bytes))
+            charBuffer.append(decoded)
+            i += 3 * bytes.length
+          } catch {
+            case _: NumberFormatException =>
+              throw new IllegalArgumentException("URLDecoder: Illegal hex characters in escape (%) pattern")
```

Comment:
Also, about the try/catch, it is customary to put one big try/catch around the loop, instead of opening/closing a try/catch on every iteration.